### PR TITLE
APDS9306 Ambient Light Sensor

### PR DESCRIPTION
**Pull request by aodrenah**
# What does this implement/fix?

Adds support for the APDS9306 ambient light sensor.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3826

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: apds9306
    name: "APDS9306 Light Level" # below are optional
    gain: 1
    bit_width: 18
    measurement_rate: 100ms
    update_interval: 60s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/apds9306/apds9306.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    return; \
+  }
+#define APDS9306_WRITE_BYTE(reg, value) \
+  ESP_LOGVV(TAG, "WRITE_BYTE: %d, %d", reg, value); \
+  APDS9306_ERROR_CHECK(this->write_byte(reg, value));
+
+void APDS9306::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up APDS9306...");
+  uint8_t id;
+  if (!this->read_byte(0x06, &id)) {  // Part ID register
```
Comment (jesserockz) [inline esphome/components/apds9306/apds9306.cpp]: Please move all of the register addresses to named constants or enums at the top of this file.

File: esphome/components/apds9306/apds9306.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  uint8_t status;
+  APDS9306_WARNING_CHECK(this->read_byte(0x07, &status), "Reading status bit failed.");
+
+  this->status_clear_warning();
+
+  if (!(status &= 0b00001000))  // No new data
+    return;
+
+  // Conversions
+  switch (gain_) {
```
Comment (jesserockz) [inline esphome/components/apds9306/apds9306.cpp]:
```suggestion
  switch (this->gain_) {
```

File: esphome/components/apds9306/apds9306.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      break;
+    case 3:
+      gain_val_ = 9;
+      break;
+    case 4:
+      gain_val_ = 18;
+      break;
+  }
+
+  switch (measurement_rate_) {
```
Comment (jesserockz) [inline esphome/components/apds9306/apds9306.cpp]:
```suggestion
  switch (this->measurement_rate_) {
```

File: esphome/components/apds9306/apds9306.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+      break;
+    case 4:
+      rate_val_ = 500;
+      break;
+    case 5:
+      rate_val_ = 1000;
+      break;
+  }
+
+  switch (bit_width_) {
```
Comment (jesserockz) [inline esphome/components/apds9306/apds9306.cpp]:
```suggestion
  switch (this->bit_width_) {
```

File: esphome/components/apds9306/apds9306.cpp [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+      break;
+    case 5:
+      bit_width_val_ = 13;
+      break;
+  }
+
+  uint8_t als_data[3];
+  APDS9306_WARNING_CHECK(this->read_byte(0x0D, als_data, 3), "Reading ALS data has failed.");
+
+  uint32_t light_level = 0x0000 | ((als_data[0]) + (als_data[1] << 8) + (als_data[2] << 16));
```
Comment (jesserockz) [inline esphome/components/apds9306/apds9306.cpp]: This is available with `#include "esphome/core/helpers.h"`
```suggestion
  uint32_t light_level = encode_uint24(als_data[2], als_data[1], als_data[0]);
```

File: esphome/components/apds9306/apds9306.cpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+  float lux = ((float) light_level / gain_val_) * (100.0f / rate_val_);
```
Comment (jesserockz) [inline esphome/components/apds9306/apds9306.cpp]:
```suggestion
  float lux = ((float) light_level / this->gain_val_) * (100.0f / this->rate_val_);
```

File: esphome/components/apds9306/sensor.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+CONF_MEASUREMENT_RATE = "measurement_rate"
+
+MEASUREMENT_BIT_WIDTHS = {
+    "20": 0,
+    "19": 1,
+    "18": 2,
+    "17": 3,
+    "16": 4,
+    "13": 5,
+}
```
Comment (jesserockz) [inline esphome/components/apds9306/sensor.py]:
```suggestion
MEASUREMENT_BIT_WIDTHS = {
    20: 0,
    19: 1,
    18: 2,
    17: 3,
    16: 4,
    13: 5,
}
```

File: esphome/components/apds9306/sensor.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    "1000ms": 5,
+}
+
+AMBIENT_LIGHT_GAINS = {
+    "1": 0,
+    "3": 1,
+    "6": 2,
+    "9": 3,
+    "18": 4,
+}
```
Comment (jesserockz) [inline esphome/components/apds9306/sensor.py]:
```suggestion
AMBIENT_LIGHT_GAINS = {
    1: 0,
    3: 1,
    6: 2,
    9: 3,
    18: 4,
}
```

File: esphome/components/apds9306/sensor.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        unit_of_measurement=UNIT_LUX,
+        accuracy_decimals=1,
+        device_class=DEVICE_CLASS_ILLUMINANCE,
+        state_class=STATE_CLASS_MEASUREMENT,
+        icon=ICON_LIGHTBULB,
+    )
+    .extend(
+        {
+            cv.Optional(CONF_GAIN, default="1"): cv.enum(
+                AMBIENT_LIGHT_GAINS, lower=True
```
Comment (jesserockz) [inline esphome/components/apds9306/sensor.py]:
```suggestion
                AMBIENT_LIGHT_GAINS, int=True
```

File: esphome/components/apds9306/sensor.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+MEASUREMENT_RATES = {
+    "25ms": 0,
+    "50ms": 1,
+    "100ms": 2,
+    "200ms": 3,
+    "500ms": 4,
+    "1000ms": 5,
+}
```
Comment (jesserockz) [inline esphome/components/apds9306/sensor.py]:
```suggestion
MEASUREMENT_RATES = {
    25: 0,
    50: 1,
    100: 2,
    200: 3,
    500: 4,
    1000: 5,
}
```

File: esphome/components/apds9306/sensor.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        state_class=STATE_CLASS_MEASUREMENT,
+        icon=ICON_LIGHTBULB,
+    )
+    .extend(
+        {
+            cv.Optional(CONF_GAIN, default="1"): cv.enum(
+                AMBIENT_LIGHT_GAINS, lower=True
+            ),
+            cv.Optional(CONF_BIT_WIDTH, default="18"): cv.enum(
+                MEASUREMENT_BIT_WIDTHS, lower=True
```
Comment (jesserockz) [inline esphome/components/apds9306/sensor.py]:
```suggestion
                MEASUREMENT_BIT_WIDTHS, int=True
```

File: esphome/components/apds9306/sensor.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+            ),
+            cv.Optional(CONF_MEASUREMENT_RATE, default="100ms"): cv.enum(
+                MEASUREMENT_RATES, lower=True
+            ),
```
Comment (jesserockz) [inline esphome/components/apds9306/sensor.py]:
```suggestion
            cv.Optional(CONF_MEASUREMENT_RATE, default="100ms"): cv.All(
                cv.positive_time_period_milliseconds, 
                cv.enum(MEASUREMENT_RATES, int=True),
            ),
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/apds9306/sensor.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import i2c, sensor
+from esphome.const import (
+    CONF_GAIN,
+    DEVICE_CLASS_ILLUMINANCE,
+    STATE_CLASS_MEASUREMENT,
+    ICON_LIGHTBULB,
+    UNIT_LUX,
```
Comment (kbx81) [inline esphome/components/apds9306/sensor.py]:
```suggestion
    CONF_GAIN,
    DEVICE_CLASS_ILLUMINANCE,
    ICON_LIGHTBULB,
    STATE_CLASS_MEASUREMENT,
    UNIT_LUX,
```

File: tests/components/apds9306/test.esp32-c3-idf.yaml
Comment (kbx81) [inline tests/components/apds9306/test.esp32-c3-idf.yaml:1]: Please consolidate the test files using a `common.yaml` file with substitutions for the pins. [Example](https://github.com/esphome/esphome/tree/dev/tests/components/ade7880)

**Review by kbx81 — DISMISSED** [review]

**Comment by aodrenah** [comment]
If you see this and I haven't commented again, please ignore this pull request for now. Since I started making changes I think I broke something  because the sensor always gets marked as failed when I test.

**Comment by kbx81** [comment]
No worries, just remember to mark as "ready" once you've fixed things up. 😄

**Comment by aodrenah** [comment]
It works again, I was using read_byte instead of read_bytes and I was resetting the device unnecessarily.

File: esphome/components/apds9306/apds9306.cpp [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+  float lux = (light_level / this->gain_val_) * (100.0f / this->measurement_time_);
+
+  ESP_LOGD(TAG, "Got illuminance=%.1flx from", lux);
+  this->publish_state(lux);
+}
+
+void APDS9306::convert_config_variables_() {
+  switch (this->gain_) {
+    case 0:
```
Comment (jesserockz) [inline esphome/components/apds9306/apds9306.cpp]: Please define and use actual enum types for all of these

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/apds9306/apds9306.cpp (hunk 14 above)
Comment (aodrenah) [reply esphome/components/apds9306/apds9306.cpp]: I switched to using an array, is this OK or did you want me to declare constants and use those inside the switch statement like I did with the registers? I'm not sure what you mean by use actual enum types.

**Comment by jesserockz** [comment]
@aodrenah I have pushed some commits to change to `enum` types where applicable.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**